File: concepts/preview-parser.cs
#!/usr/bin/env -S dotnet --
#:property LangVersion=preview

using System.Text.RegularExpressions;

string input = "int main(void) { return 2; }";

List<Token> tokens = Lex(input);

C.Program program = ParseProgram(tokens);

Console.WriteLine(program);

List<Token> Lex(string input)
{
    List<Token> tokens = [];

    while (input != string.Empty)
    {
        input = input.TrimStart();
        if (input == string.Empty)
            break;

        bool matched = false;

        foreach ((Type type, Regex pattern) in Patterns)
        {
            Match match = pattern.Match(input);
            if (!match.Success || match.Index != 0)
                continue;

            tokens.Add(CreateToken(type, match.Value));
            input = input[match.Length..];
            matched = true;
            break;
        }

        if (!matched)
            throw new InvalidOperationException($"Unexpected character: '{input[0]}'");
    }

    return tokens;
}

Token CreateToken(Type type, string value) => type switch
{
    var _ when type == typeof(Identifier) => new Identifier(value),
    var _ when type == typeof(Int) => new Int(),
    var _ when type == typeof(Void) => new Void(),
    var _ when type == typeof(Return) => new Return(),
    var _ when type == typeof(Constant) => new Constant(int.Parse(value)),
    var _ when type == typeof(OpenParenthesis) => new OpenParenthesis(),
    var _ when type == typeof(CloseParenthesis) => new CloseParenthesis(),
    var _ when type == typeof(OpenBrace) => new OpenBrace(),
    var _ when type == typeof(CloseBrace) => new CloseBrace(),
    var _ when type == typeof(Semicolon) => new Semicolon(),
    _ => throw new InvalidOperationException()
};

C.Program ParseProgram(List<Token> tokens)
{
    return new C.Program(ParseFunction(tokens));
}

C.Function ParseFunction(List<Token> tokens)
{
    Expect<Int>(tokens);
    string identifier = ParseIdentifier(tokens);
    Expect<OpenParenthesis>(tokens);
    Expect<Void>(tokens);
    Expect<CloseParenthesis>(tokens);
    Expect<OpenBrace>(tokens);
    C.Return statement = ParseStatement(tokens);
    Expect<CloseBrace>(tokens);

    return new C.Function(identifier, statement);
}

C.Return ParseStatement(List<Token> tokens)
{
    Expect<Return>(tokens);
    C.Constant expression = ParseExpression(tokens);
    Expect<Semicolon>(tokens);

    return new C.Return(expression);
}

C.Constant ParseExpression(List<Token> tokens)
{
    return new C.Constant(ParseInt(tokens));
}

string ParseIdentifier(List<Token> tokens)
{
    if (TakeToken(tokens) is Identifier identifier)
        return identifier.Value;

    throw new InvalidOperationException();
}

int ParseInt(List<Token> tokens)
{
    if (TakeToken(tokens) is Constant constant)
        return constant.Value;

    throw new InvalidOperationException();
}

void Expect<T>(List<Token> tokens)
{
    if (TakeToken(tokens) is not T)
        throw new InvalidOperationException();
}

Token TakeToken(List<Token> tokens)
{
    Token token = tokens.First();
    tokens.RemoveAt(0);
    return token;
}

readonly union Token
(
    Identifier,
    Int,
    Void,
    Return,
    Constant,
    OpenParenthesis,
    CloseParenthesis,
    OpenBrace,
    CloseBrace,
    Semicolon
);

record Identifier(string Value);
record Int();
record Void();
record Return();
record Constant(int Value);
record OpenParenthesis();
record CloseParenthesis();
record OpenBrace();
record CloseBrace();
record Semicolon();

namespace C
{
    record Program(Function Function);
    record Function(string Identifier, Return Statement);
    record Return(Constant Expression);
    record Constant(int Value);
}

namespace Assembly
{
    record Program(Function Function);
    record Function(string Identifer, List<Instruction> Instructions);
    readonly union Instruction(Mov, Ret);
    record Mov(Operand Source, Operand Destination);
    record Ret();
    readonly union Operand(Imm, Register);
    record Imm(int Value);
    record Register();
}

partial class Program
{
    private static readonly List<KeyValuePair<Type, Regex>> Patterns =
    [
        new(typeof(Int), IntPattern),
        new(typeof(Void), VoidPattern),
        new(typeof(Return), ReturnPattern),
        new(typeof(Identifier), IdentifierPattern),
        new(typeof(Constant), ConstantPattern),
        new(typeof(OpenParenthesis), OpenParenthesisPattern),
        new(typeof(CloseParenthesis), CloseParenthesisPattern),
        new(typeof(OpenBrace), OpenBracePattern),
        new(typeof(CloseBrace), CloseBracePattern),
        new(typeof(Semicolon), SemicolonPattern)
    ];

    [GeneratedRegex(@"[a-zA-Z_]\w*\b")]
    private static partial Regex IdentifierPattern { get; }

    [GeneratedRegex(@"\bint\b")]
    private static partial Regex IntPattern { get; }

    [GeneratedRegex(@"\bvoid\b")]
    private static partial Regex VoidPattern { get; }

    [GeneratedRegex(@"\breturn\b")]
    private static partial Regex ReturnPattern { get; }

    [GeneratedRegex(@"[0-9]+\b")]
    private static partial Regex ConstantPattern { get; }

    [GeneratedRegex(@"\(")]
    private static partial Regex OpenParenthesisPattern { get; }

    [GeneratedRegex(@"\)")]
    private static partial Regex CloseParenthesisPattern { get; }

    [GeneratedRegex(@"{")]
    private static partial Regex OpenBracePattern { get; }

    [GeneratedRegex(@"}")]
    private static partial Regex CloseBracePattern { get; }

    [GeneratedRegex(@";")]
    private static partial Regex SemicolonPattern { get; }
}

File: concepts/preview-parser.rs
#[derive(Clone, Copy, PartialEq)]
enum Token<'a> {
    Identifier(&'a str),
    Int,
    Void,
    Return,
    Constant(u32),
    OpenParenthesis,
    CloseParenthesis,
    OpenBrace,
    CloseBrace,
    Semicolon,
}

struct Program<'a> {
    function: Function<'a>,
}

struct Function<'a> {
    identifier: &'a str,
    statement: Return,
}

struct Return {
    expression: Constant,
}

struct Constant {
    value: u32,
}

fn main() {
    let tokens = lex("int main(void) { return 2; }").expect("lex failed");
    let mut parser = Parser::new(&tokens);
    let program = parser.parse_program().expect("parse failed");

    assert_eq!(program.function.identifier, "main");
    assert_eq!(program.function.statement.expression.value, 2);
    println!(
        "Program {{ Function = Function {{ Identifier = {}, Statement = Return {{ Expression = Constant {{ Value = {} }} }} }} }}",
        program.function.identifier, program.function.statement.expression.value
    );
}

fn lex(input: &str) -> Result<Vec<Token<'_>>, &'static str> {
    let bytes = input.as_bytes();
    let mut tokens = Vec::new();
    let mut index = 0;

    while index < bytes.len() {
        if bytes[index].is_ascii_whitespace() {
            index += 1;
            continue;
        }

        if bytes[index].is_ascii_alphabetic() || bytes[index] == b'_' {
            let start = index;
            index += 1;
            while index < bytes.len()
                && (bytes[index].is_ascii_alphanumeric() || bytes[index] == b'_')
            {
                index += 1;
            }

            tokens.push(match &input[start..index] {
                "int" => Token::Int,
                "void" => Token::Void,
                "return" => Token::Return,
                identifier => Token::Identifier(identifier),
            });
            continue;
        }

        if bytes[index].is_ascii_digit() {
            let start = index;
            while index < bytes.len() && bytes[index].is_ascii_digit() {
                index += 1;
            }
            let value = input[start..index]
                .parse()
                .map_err(|_| "invalid integer")?;
            tokens.push(Token::Constant(value));
            continue;
        }

        tokens.push(match bytes[index] {
            b'(' => Token::OpenParenthesis,
            b')' => Token::CloseParenthesis,
            b'{' => Token::OpenBrace,
            b'}' => Token::CloseBrace,
            b';' => Token::Semicolon,
            _ => return Err("unexpected character"),
        });
        index += 1;
    }

    Ok(tokens)
}

struct Parser<'a> {
    tokens: &'a [Token<'a>],
    current: usize,
}

impl<'a> Parser<'a> {
    fn new(tokens: &'a [Token<'a>]) -> Self {
        Self { tokens, current: 0 }
    }

    fn parse_program(&mut self) -> Result<Program<'a>, &'static str> {
        let program = Program {
            function: self.parse_function()?,
        };
        if self.current != self.tokens.len() {
            return Err("unexpected trailing token");
        }
        Ok(program)
    }

    fn parse_function(&mut self) -> Result<Function<'a>, &'static str> {
        self.expect(Token::Int)?;
        let identifier = self.parse_identifier()?;
        self.expect(Token::OpenParenthesis)?;
        self.expect(Token::Void)?;
        self.expect(Token::CloseParenthesis)?;
        self.expect(Token::OpenBrace)?;
        let statement = self.parse_statement()?;
        self.expect(Token::CloseBrace)?;
        Ok(Function {
            identifier,
            statement,
        })
    }

    fn parse_statement(&mut self) -> Result<Return, &'static str> {
        self.expect(Token::Return)?;
        let expression = self.parse_expression()?;
        self.expect(Token::Semicolon)?;
        Ok(Return { expression })
    }

    fn parse_expression(&mut self) -> Result<Constant, &'static str> {
        Ok(Constant {
            value: self.parse_integer()?,
        })
    }

    fn parse_identifier(&mut self) -> Result<&'a str, &'static str> {
        match self.take()? {
            Token::Identifier(identifier) => Ok(identifier),
            _ => Err("expected identifier"),
        }
    }

    fn parse_integer(&mut self) -> Result<u32, &'static str> {
        match self.take()? {
            Token::Constant(constant) => Ok(constant),
            _ => Err("expected integer"),
        }
    }

    fn expect(&mut self, expected: Token<'a>) -> Result<(), &'static str> {
        if self.take()? != expected {
            return Err("unexpected token");
        }
        Ok(())
    }

    fn take(&mut self) -> Result<Token<'a>, &'static str> {
        let token = self
            .tokens
            .get(self.current)
            .copied()
            .ok_or("unexpected end of input")?;
        self.current += 1;
        Ok(token)
    }
}
